### ISSUE TITLE
Feature: MPI support for Tomopy using streams

### ISSUE THREAD

**Issue by michael-sutherland** [tomopy/tomopy#179]
I thought I'd start a discussion on MPI support for Tomopy.  I have a very simplistic example for using MPI on my [fork of Tomopy](https://github.com/michael-sutherland/tomopy/tree/mpi), I only tested normalization and recon functions (see example at end).  We really want to avoid sending data back and forth between processes if possible; it didn't scale and was slow, even on a single computer.  

So the next idea (thanks Doga) is to use a streaming interface for Tomopy.  The general idea would be to set up a pipeline of processing: read from file -> normalize -> remove stripes-> extract sinograms -> reconstruct -> write to file.  Hopefully, we could then optimize the data handling so that multiple operations could be performed on a chunk of data before moving to the next chunk.  For example, normalize data as it is being read in from file.  Or do all operations in sinogram order if possible, so that we can stream sinograms from file to processing to reconstruction and back to file.  A streaming interface should expand flexibility and reduce RAM usage.

I'm hoping a streaming interface would look something like the following:

```
reader = tomopy.project_specific_reader(filename, options) # lazy load data from file (could be distributed)
normalize = tomopy.normalize(reader, options)
stripes_removed = tomopy.remove_stripes(normalize, options)
recon = tomopy.recon(stripes_removed, options) # data would be rotated here
output = tomopy.write_file(recon, options) # writing to file is were processing actually kicks off streams
```

You should be able to get the dataset at any point for custom processing or debugging:

```
reader = tomopy.project_specific_reader(filename, options) # lazy load data from file (could be distributed)
normalize = tomopy.normalize(reader, options)
data = tomopy.get_data(normalize) # would process everything up to this point, should be able to continue here or with normalize.
stripes_removed = tomopy.remove_stripes(normalize, options) # might be tricky to support multiple outputs for streams, especially if we want to avoid copies.
...
```

Under the hood, iterators would be used to manage the streams with a bit of meta-data for chunk size, order of the data (radiograph or sonogram), overlap, etc.  It might be easiest to use a scheduling/management object to manage the data passing, which could provide overall limits on things like memory consumption.

I won't have time to work on this much in the next couple of weeks, but want to take a stab at it at some point.  If anyone has any other ideas (especially something simpler), I'd love to hear it.

Here's some example code to use the current MPI repo:

```
from __future__ import print_function, division

from mpi4py import MPI 
import numpy as np
import os
import time
import tomopy

import logging
import math

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

#NOTE: basicConfig ignored when using IPython
if rank == 0:
    logging.basicConfig(#filename='log.log',
                        format='%(asctime)s - %(levelname)s - %(name)s - %(funcName)s - %(message)s',
                        level=logging.INFO)
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

working_dir = os.path.join('.')
filename = 'filename.h5'
rot_center = 1281
theta_start = 20
theta_end = 160

# Main Code
# ------------------------------------------------------------

start = time.time()
if rank == 0: 
    logger.info("Start")    

# change to working dir
os.chdir(working_dir)

# %% Read HDF5 file.
if rank == 0: 
    logger.info("Reading data from H5 file %s" % filename)
#TODO: read directly into different mpi processes
if rank == 0:
    # read data into root node
    proj, flat, dark = tomopy.read_aps_32id(filename, sino=slice(512,-512),dtype=np.float32, shared=True)
else:
    proj, flat, dark = None, None, None


# %% Flat field correct data
if rank == 0: 
    logger.info("Flat field correcting data")
proj = tomopy.normalize(proj, flat, dark)
del flat, dark

# %% Generate theta
if rank == 0: 
    logger.info("Generating theta")
    shape = proj.shape
else:
    shape = None
shape = comm.bcast(shape)
theta = np.linspace(theta_start, theta_end, shape[0], endpoint=True)

if rank == 0: 
    np.clip(proj, 1e-6, 1.0, proj)
    logger.info("Rotate dataset so that y is first and theta second.")
    rot_start = time.time()
    proj = tomopy.init_tomo(proj, False, False, False)
    logger.info("Rotation took %0.2f seconds" % (time.time() - rot_start))

alg = 'gridrec'
if rank == 0: 
    logger.info("Reconstructing using: %s" % alg)

# Reconstruct object using algorithm
start_recon = time.time()
rec = tomopy.recon(proj,
                   theta*(math.pi/180.0),
                   center=rot_center,
                   algorithm=alg,                           
                   nchunk=2,
                   emission=True,
                   sinogram_order=True)
if rank == 0: 
    logger.info("Recon %s took %0.2f seconds" % (alg, time.time() - start_recon))


# Write data as stack of TIFs.
if rank == 0:
    tomopy.write_tiff_stack(rec, fname='%s(mpi,c=%0.1f)/recon' % (alg, rot_center), overwrite=True)
del rec

if rank == 0: 
    logger.info("Done in %0.2f seconds"%(time.time() - start))

```

**Comment by michael-sutherland** [comment]
After having a meeting at ALS to discuss Tomopy running on a cluster, we are going to attempt the following changes:

1. @dmpelt is planning to move everything in tomopy to multithreading instead of multprocessing.  This should address the various issues people have had with multiprocessing and remove all shared memory code.  I believe this will allow us to bring back native Windows support (see issue #110).  This should still be fast (and potentially faster) if we make sure our code is releasing the GIL (see pull request #217).

2. I am going to work on adding MPI support on top of Tomopy.  We want Tomopy to be able to run efficiently on a workstation without MPI and on a cluster.  We considered using Dask or Spark as a distribution tool, but decided to start with MPI.  Dask and Spark both seem to have memory consumption issues and make a lot of copies of the data, although it could just be we need someone with better experience to show us what we are missing.

3. We want to improve dxchange to support meta data processing and lazy loading of image data.  I am thinking of adding support for Pandas for Meta-Data (allowing each project to add their own custom fields).  Hopefully we can keep a common set of names for common parameters like acqusition timestamp, camera exposure time, theta, etc.

**Comment by dmpelt** [comment]
Point 1 is going well (see [here](https://github.com/dmpelt/tomopy/tree/threading)). The two remaining routines to convert are `remove_stripe_fw` and `remove_stripe_ti`, which are both Python-heavy. I am having a bit of trouble getting them to perform well with threading, but I think it is doable for `remove_stripe_fw`. 

`remove_stripe_ti` is more difficult, since it has many small Python calls, so maybe it has to be rewritten (it looks like it can be optimized quite a bit anyway). For now, we can do three things I think:
* keep `multiprocessing` only for `remove_stripe_ti`
* use `concurrent.futures` for `remove_stripe_ti`, which will make it slow
* let `remove_stripe_ti` run only on a single core, which will make it slow

What do you think is the best option for now?

**Comment by xianghuix** [comment]
I vote for third option. This option will not cost much effort and the performance may not be very different from option 2. We can live with slow operation without hanging until someone can find a better solution.


________________________________

**Comment by decarlof** [comment]
Thank you all for taking the lead of this. Here are my 2 cents:
point 1. keep the code readable, this makes it easy to contribute to (I have seen highly optimized code that are very difficult to understand). 
point 2. make sure non-MPI users don't see the difference. I don't know mpi but I think building MPI support on top already means this = the MPI calls are made to tomopy code with no need to incorporate any MPI code into tomopy, correct?
point 3. I am very supportive of adding meta data to handle processing provenance and Panda seems a very good tool for this. From the hdf5 implementation of data exchange for tomography (see [dxchange](http://dxfile.readthedocs.io) for details) we proposed a ["process"](http://dxfile.readthedocs.io/en/latest/source/xraytomo.html#process) tag under which to store a processing table events. At the moment its content is quite open to changes and redefinition since besides the acquisition and basic recon parameters for gridrec is basically not used.

**Comment by michael-sutherland** [comment]
@dmpelt, I vote for option 2 (a user could specify ncore = 1 to run single threaded).  We can then try to optimize the code a bit for running multithreaded as well.  We could probably get rid of a few of the data copies that currently exist (add sinogram_order as an option to all remove_stripe functions).

**Comment by michael-sutherland** [comment]
@decarlof Good points.  Yes, that was the intention for point 2.  I might need to add some things to tomopy to help with running MPI, but it should be completely transparent to the user.  Specifically, I'm thinking of documenting which axis the data is being split on for the various functions (maybe we could use a decorator?). 

For point 3, I really see the need to be able to get meta-data from a dataset without reading in the actual dataset.  For example, it would be extremely useful to get the size of a dataset without reading it in.  If we could separate the getting of meta-data from reading data, we could programmatically schedule the work that needs to be done based on the resources available (memory limitations, workstation or cluster, etc).

**Comment by decarlof** [comment]
@michael-sutherland for point 3: if the file is hdf you can read the dataset meta data without reading the actual data with:

f=h5py.File("myhdf.h5", "r")
data = f['exchange/data']
print data
<HDF5 dataset "data": shape (1500, 1200, 1920), type "<u2">
print data.shape
(1500, 1200, 1920)

**Comment by michael-sutherland** [comment]
Quick update: I've tried to use distarray and have run into a few different issues.  It appears the library uses pickle to distribute data to and from the remote processes.  This is very inefficient and can't handle very large arrays (over 2GB in size).  It also appears distarray as a project has been abandoned.  I really like some of the features, but it is looking like I will need to create my own MPI distribution code for now.  I'll update when I get something working well.

**Comment by michael-sutherland** [comment]
pywt is planning on releasing a new version (0.5.0) with @dmpelt 's changes that should free up the GIL (See https://github.com/PyWavelets/pywt/issues/203).  Switching to pywt 0.5.0 should speed up remove_strip_fw using threads instead of multiprocessing.  Also, it seems that using multiprocessing can cause issues when running with MPI, so we'll get an added benefit from this change.

**Comment by dmpelt** [comment]
I have a branch of tomopy [here](https://github.com/dmpelt/tomopy/tree/threading) with a threaded version of `remove_stripe_fw`. It turned out to be quite difficult to get it to perform efficiently, so the code was rewritten quite a bit. I still have to document it and maybe restructure it to read better, but if you want you can already test it. It indeed requires the new version of pywt (or the current master).

**Comment by michael-sutherland** [comment]
Neat!  I also see the other functions that need to be switched from multiprocessing to multithreading in your fork.  I'm hoping we can completely switch to multithreading soon!

I'd also suggest adding an option for sinogram_order to the remove_strip_fw function.  It appears the function actually only operates on a single slice across axis 1 at a time (which is also interesting, I think the algorithm could be updated to include influence from the neihboring sinogram slices).  When doing distributed processing, we'll have to pass in data in sinogram order.

Along those lines, [here](https://github.com/michael-sutherland/tomopy/tree/mpiarray) is the mpiarray python object I've been working on to support clustering with Tomopy.  I'm hoping we can try it out soon on Cori or Edison.  We'll need to make some changes to Tomopy (like multithreading support) to really get this working well.

**Comment by wscullin** [comment]
@michael-sutherland I somehow missed the discussion from November on. What sort of data are you moving around? If it's numpy arrays, mpi4py should be able to move them without pickling though the API is a bit different.

**Comment by michael-sutherland** [comment]
@wscullin It is numpy arrays.  I created mpiarray to manage the passing around of numpy arrays using mpi4py.  It gets tricky if you want to split on an axis that isn't zero, redistribute data, support padding, or do a swapaxes to switch from a stack of images to a stack of sinograms.  I've got a new version of mpiarray that supports these functions that I just need to host somewhere.

Also, there are a number of updates we need to do to Tomopy to follow up on supporting MPI.  Mainly, removing any forking (as mentioned above).  Also, there was an effort at NERSC to study using MPI with and without multithreading with Tomopy, but I haven't seen the results yet.
